{
 "cells": [
  {
   "cell_type": "markdown",
   "id": "abececa3",
   "metadata": {},
   "source": [
    "# Track data lineage"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 1,
   "id": "d24726db",
   "metadata": {},
   "outputs": [
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "âœ… Loaded instance: lawrence-lamin/lamindb (lamindb 0.43.0)\n"
     ]
    }
   ],
   "source": [
    "import lamindb as ln"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 2,
   "id": "08383c6c",
   "metadata": {},
   "outputs": [
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "ðŸ’¬ Notebook imports: lamindb==0.43.0\n",
      "ðŸ”¶ Automatic tracking of notebook failed: insert or update on table \"lnschema_core_transform\" violates foreign key constraint \"lnschema_core_transf_created_by_id_8afc7db3_fk_lnschema_\"\n",
      "DETAIL:  Key (created_by_id)=(325XQ1LU) is not present in table \"lnschema_core_user\".\n",
      "\n",
      "ðŸ”¶ No automatic metadata detection, consider passing transform\n"
     ]
    }
   ],
   "source": [
    "ln.track()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 4,
   "id": "b5bacde8",
   "metadata": {},
   "outputs": [
    {
     "ename": "RuntimeError",
     "evalue": "Your DB is not initialized: lnschema_core_user has no row",
     "output_type": "error",
     "traceback": [
      "\u001b[0;31m---------------------------------------------------------------------------\u001b[0m",
      "\u001b[0;31mRuntimeError\u001b[0m                              Traceback (most recent call last)",
      "Cell \u001b[0;32mIn[4], line 1\u001b[0m\n\u001b[0;32m----> 1\u001b[0m \u001b[43mln\u001b[49m\u001b[38;5;241;43m.\u001b[39;49m\u001b[43msetup\u001b[49m\u001b[38;5;241;43m.\u001b[39;49m\u001b[43mload\u001b[49m\u001b[43m(\u001b[49m\u001b[38;5;124;43m\"\u001b[39;49m\u001b[38;5;124;43mlamindb\u001b[39;49m\u001b[38;5;124;43m\"\u001b[39;49m\u001b[43m)\u001b[49m\n",
      "File \u001b[0;32m/opt/conda/lib/python3.10/site-packages/lamindb_setup/_load_instance.py:76\u001b[0m, in \u001b[0;36mload\u001b[0;34m(identifier, storage, _log_error_message, _access_token, _test)\u001b[0m\n\u001b[1;32m     74\u001b[0m \u001b[38;5;28;01mif\u001b[39;00m \u001b[38;5;129;01mnot\u001b[39;00m check:\n\u001b[1;32m     75\u001b[0m     \u001b[38;5;28;01mif\u001b[39;00m _log_error_message:\n\u001b[0;32m---> 76\u001b[0m         \u001b[38;5;28;01mraise\u001b[39;00m \u001b[38;5;167;01mRuntimeError\u001b[39;00m(msg)\n\u001b[1;32m     77\u001b[0m     \u001b[38;5;28;01melse\u001b[39;00m:\n\u001b[1;32m     78\u001b[0m         logger\u001b[38;5;241m.\u001b[39mwarning(\n\u001b[1;32m     79\u001b[0m             \u001b[38;5;124m\"\u001b[39m\u001b[38;5;124mInstance metadata exists, but DB might have been corrupted or deleted.\u001b[39m\u001b[38;5;124m\"\u001b[39m\n\u001b[1;32m     80\u001b[0m             \u001b[38;5;124m\"\u001b[39m\u001b[38;5;124m Re-initializing the DB.\u001b[39m\u001b[38;5;124m\"\u001b[39m\n\u001b[1;32m     81\u001b[0m         )\n",
      "\u001b[0;31mRuntimeError\u001b[0m: Your DB is not initialized: lnschema_core_user has no row"
     ]
    }
   ],
   "source": [
    "ln.setup.load(\"lamindb\")"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 5,
   "id": "5ebd8afd-a337-4cd3-86a4-7dffb8894712",
   "metadata": {},
   "outputs": [
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "ðŸ”¶ Instance metadata exists, but DB might have been corrupted or deleted. Re-initializing the DB.\n",
      "âœ… Registered instance on hub: https://lamin.ai/lawrence-lamin/lamindb\n",
      "ðŸ’¬ Creating schemas: core==0.35.9 bionty==0.22.0 \n",
      "âœ… Saved: User(id=325XQ1LU, handle=lawrence-lamin, email=lawrlee+lamin@gmail.com, name=Dr Lawrence Lee, created_at=2023-06-16 17:22:30.637067, updated_at=2023-06-16 17:22:30.637294)\n",
      "âœ… Saved: Storage(id=q9t77OHI, root=s3://lamin-lawrence, type=s3, region=us-east-1, created_at=2023-06-16 17:22:30.690630, updated_at=2023-06-16 17:22:30.690639, created_by_id=325XQ1LU)\n",
      "âœ… Loaded instance: lawrence-lamin/lamindb (lamindb 0.43.0)\n"
     ]
    }
   ],
   "source": [
    "ln.setup.init(\n",
    "    storage=\"s3://lamin-lawrence\", \n",
    "    db=\"postgresql://lamin:lamin_pw@postgres:5432/lamindb\",\n",
    "    schema=\"bionty\"\n",
    ")"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 6,
   "id": "617f0040-f0b1-4591-9987-86a940f321e9",
   "metadata": {},
   "outputs": [
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "ðŸ’¬ Notebook imports: lamindb==0.43.0\n",
      "âœ… Saved: Transform(id=p3MdDgPklTBecI, name=Track data lineage, short_name=lamin_data_2, stem_id=p3MdDgPklTBe, version=0, type=notebook, reference=None, created_at=2023-06-16 17:22:55.184857, updated_at=2023-06-16 17:22:55.184864, created_by_id=325XQ1LU)\n",
      "âœ… Saved: Run(id=nPCvVzbvHaTQXh2BLjVA, name=None, external_id=None, created_at=2023-06-16 17:22:55.191098, run_at=2023-06-16 17:22:55.191108, transform_id=p3MdDgPklTBecI, created_by_id=325XQ1LU)\n"
     ]
    }
   ],
   "source": [
    "ln.track()"
   ]
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "Python 3 (ipykernel)",
   "language": "python",
   "name": "python3"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 3
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython3",
   "version": "3.10.11"
  },
  "nbproject": {
   "id": "p3MdDgPklTBe",
   "parent": null,
   "pypackage": null,
   "time_init": "2023-06-16T01:05:48.448020+00:00",
   "user_handle": null,
   "user_id": null,
   "user_name": null,
   "version": "0"
  }
 },
 "nbformat": 4,
 "nbformat_minor": 5
}
